### PR TITLE
Codec adapter fixups

### PR DESCRIPTION
**Pull request by cujomalainey**
For the documentation fix I personally don't agree with the design but the documentation needs to match the code. **Begin rant** I think the adapter should **never** reject a config push to a valid adapter. This makes it hard for things like audio servers to push config's on boot as its past init so too late for setup but too early for runtime. The library should be allowed to make the choice when it wants to use the config but it should always be notified of when it is loaded. Also I don't think we should be limiting the types, why not make the id generic and let the adapter decide what types of configs it wants and pass that as an argument in the callback? Right now the cadence adapter uses the same function for both setup and runtime config which suggests that we should be something similar to the command workflow in components rather than creating a diamond code path unnecessarily. **end rant**

I am hesitant to change this API as I know it needs to be stable but these issues worry me as they appear to already be causing confusion. I am happy to fork this into a issue and continue this discussion if need be there.

### CODE REVIEW

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
@cujomalainey I think this is a really good topic for January TSC (if not resolved before hand). We should continue the discussion in GH in the mean time. I will create a new TSC meeting for end of January. I think we should be able to get in updates in v1.8

File: src/include/sof/audio/codec_adapter/codec/generic.h [hunk 1]
```diff
@@ -57,7 +57,8 @@ struct codec_interface {
 	int (*process)(struct comp_dev *dev);
 	/**
 	 * Codec specific apply config procedure, called by codec_adapter every time
-	 * new configuration has been loaded.
+	 * a new RUNTIME configuration has been if in correct state. This will not
```
Comment (mrajwa) [inline src/include/sof/audio/codec_adapter/codec/generic.h:60]: The proposed sentence doesn't seem to be correct, what about
"...a new RUNTIME configuration has been sent, assuming codec is already prepared. This will not be used for SETUP config."

**Comment by mrajwa** [comment]
@cujomalainey thanks for your feedback. I am not sure if I got the problem you are facing. Let's discuss it:

> I think the adapter should never reject a config push to a valid adapter
Well, why would you push RUNTIME config to a codec which is not initialized nor prepared? What should such codec do with your runtime config? To me the only logical choice is to stash such config in the memory and wait until codec gets prepared - and this is exactly what we do with only on difference - we keep that config on codec_adapter (higher) level, which I believe is fine. 
Currently you can sent runtime parameters as soon as codec_adapter gets created so I don't see any problem here. Would you like to be able to send these parameters before codec_adapter is created?

> This makes it hard for things like audio servers to push config's on boot as its past init so too late for setup but too early for runtime.

Please define "boot" and "init" here.

> why not make the id generic and let the adapter decide what types of configs it wants and pass that as an argument in the callback?

The ID is generic, it the header of each config you specify what type of config is it. Codec adapter checks what config type it was given and reacts accordingly. 

>   Right now the cadence adapter uses the same function for both setup and runtime config which suggests that we should be something similar to the command workflow in components rather than creating a diamond code path unnecessarily

Its true that cadence does use the same function to apply both setup & runtime configs because cadence formats both of them in the same way. However it doesn't mean we should code higher layer in this way for at least two reasons:
 - on the SoF component layer its better to have two distinct types of configurations so you can operate on them base on component state, here again - why would we pass RUNTIME config to a codec which is not prepared yet?
 - other codecs may treat setup and runtime parameters differently

To summarize I don't see an issue at the moment.

**Comment by cujomalainey** [comment]
> @cujomalainey thanks for your feedback. I am not sure if I got the problem you are facing. Let's discuss it:
> 
> > I think the adapter should never reject a config push to a valid adapter
> 
> Well, why would you push RUNTIME config to a codec which is not initialized nor prepared? What should such a codec do with your runtime config? To me the only logical choice is to stash such config in the memory and wait until prepare - and this is exactly what we do with only one difference - we keep that config on codec_adapter (higher) level, which I believe is sane.
> Currently you can sent runtime parameters as soon as codec_adapter gets created so I don't see any problem here. Would you like to be able to send these parameters before codec_adapter is created?
> 

The codec can make the decision if its ready for the config, otherwise pick it up at the prepare step, but the callback should be called regardless for the codec to make that decision. Some codecs/libraries may not need a setup config or even need to be prepared to have their config set so this just gets in the way of that workflow. I have had external teams think their config push is failing because they are not seeing the callback (partially docs failure here) but also misleading code. In my mind an adapter is supposed to provide a new interface without any managerial logic which is happening here.

> > This makes it hard for things like audio servers to push config's on boot as its past init so too late for setup but too early for runtime.
> 
> Please define "boot" and "init" here.

On Chromeos we typically set configs via UCM. init We can't load setup via topology since we use the same topology possibly for different builds. Teams are setting runtime params via ucm and thinking their code is failling because nothing has been prepared even though they don't care. They don't want to use device init since the config will remain the same and would just slow the device open step down. The current restriction makes static configs much more annoyig to set once and forget.
> 
> > why not make the id generic and let the adapter decide what types of configs it wants and pass that as an argument in the callback?
> 
> The ID is generic, in the header of each config you specify what type is it. Codec adapter checks what config type it was given and reacts accordingly.

No its not generic if the codec adapter is intervening. By generic I mean the adapter has no understanding of what the IDs represent and the codec handles and defines them. Then each codec can have its own range of config fields.
> 
> > Right now the cadence adapter uses the same function for both setup and runtime config which suggests that we should be something similar to the command workflow in components rather than creating a diamond code path unnecessarily
> 
> Its true that cadence does use the same function to apply both setup & runtime configs because cadence formats both of them in the same way. However it doesn't mean we should code higher layer similarly for at least two reasons:
> 
> * on the SoF component layer its better to have two distinct types of configurations so you can operate on them base on component state, here again - why would we pass RUNTIME config to a codec which is not prepared yet?
> * other codecs may treat setup and runtime parameters differently

Again you are relying on an assumption that all codecs will operate the same and I think that is dangerous for a long term stable API. For SOF components all config handling is sent through a single path regardless of state (except when the DSP is physically asleep), I am arguing we should be doing the same here. Passing all config calls to a single endpoint and then branching out rather than figuring out the call based on where it came from which is what you have in the cadence code currently. State is fine to be maintained by the adapter but I do not think it should inhibit the codec from acting on anything the developer desires, that is up to the codec to decide.

> 
> To summarize I don't see an issue at the moment.

**Comment by mrajwa** [comment]
>  I have had external teams think their config push is failing because they are not seeing the callback (partially docs failure here) but also misleading code.

Hmm the SOF logs should clearly say what was the issue with callback. If their codec doesn't have any runtime params then there is no problem. On the other hand SETUP config is part of component creation - it was designed this way for convenience, you have to send the setup parameters at some point anyway so it is more convenient to couple them together with SoF component params otherwise you would need to send them separately. 

> On Chromeos we typically set configs via UCM. init We can't load setup via topology since we use the same topology possibly for different builds.

I don't know UCM, but it probably sends just separate IPC with configs. We can rework the part of codec adapter which forces to send SETUP config together with SoF component config if that is problematic.

> Again you are relying on an assumption that all codecs will operate the same and I think that is dangerous for a long term stable API.

Codec_adpater just provides a fremowork and therefore enforces some constrains on codecs so they work in predictable, safe manner. On the other hand it should be flexible enough to allow any codec integration. Sure we can relax some of the constrains, the setup params case is just a good example here.

**Comment by cujomalainey** [comment]
Apologies for delayed response, had to focus internally for a bit

> > I have had external teams think their config push is failing because they are not seeing the callback (partially docs failure here) but also misleading code.
> 
> Hmm the SOF logs should clearly say what was the issue with callback. If their codec doesn't have any runtime params then there is no problem. On the other hand SETUP config is part of component creation - it was designed this way for convenience, you have to send the setup parameters at some point anyway so it is more convenient to couple them together with SoF component params otherwise you would need to send them separately.

I don't think that is true, you appear to have a double check, one in [generic](https://github.com/thesofproject/sof/blob/master/src/audio/codec_adapter/codec/generic.c#L285) and another [codec_adapter](https://github.com/thesofproject/sof/blob/master/src/audio/codec_adapter/codec_adapter.c#L537) and the codec_adapter check does not emit a message and comes first.

As for setup this means that for setup to be effective it has to be coupled with the topology which seems to make the point of being able to route amixer controls kind of backwards.
> 
> > On Chromeos we typically set configs via UCM. init We can't load setup via topology since we use the same topology possibly for different builds.
> 
> I don't know UCM, but it probably sends just separate IPC with configs. We can rework the part of codec adapter which forces to send SETUP config together with SoF component config if that is problematic.

UCM is not a matter of how, but a matter of when. UCM is applied by the userspace audio server and contains a lot of "setup" style data.  The issue is that is after the probe of the DSP and loading of the topology. UCM's are an excellent way of configuring different devices without modifying kernel paths or the topology itself. I don't think we should inhibit this configuration path. I think IIR and FIR components are an excellent example where they can take a config at anytime and just pick it up when its time to use it.
> 
> > Again you are relying on an assumption that all codecs will operate the same and I think that is dangerous for a long term stable API.
> 
> Codec_adpater just provides a fremowork and therefore enforces some constrains on codecs so they work in predictable, safe manner. On the other hand it should be flexible enough to allow any codec integration. Sure we can relax some of the constrains, the setup params case is just a good example here.
> 
> @cujomalainey do you have any change request ready so we could review it?

I think we should relax all constraints, unless we have something that SOF itself requires us to constrain, to allow best interoperability. I will hopefully have some more changes in the next week or two.

File: src/include/sof/audio/codec_adapter/codec/generic.h (hunk 1 above)
Comment (cujomalainey) [reply src/include/sof/audio/codec_adapter/codec/generic.h:60]: yep, i think i missed a word, thanks

**Comment by cujomalainey** [comment]
Another point raised, why do we have a blob size restriction in the adapter? I thought we have size restriction tools in the topology, shouldnt we use that rather than create duplicate code?

https://github.com/thesofproject/sof/commit/1ac1adb5b1a0e7f942216d26b471547e1dae1de8

**Comment by lgirdwood** [comment]
> Another point raised, why do we have a blob size restriction in the adapter? I thought we have size restriction tools in the topology, shouldnt we use that rather than create duplicate code?
> 
> [1ac1adb](https://github.com/thesofproject/sof/commit/1ac1adb5b1a0e7f942216d26b471547e1dae1de8)

Can't think of any FW/SW reason, could be a limitation due to available memory in the initial implementation target ?

**Comment by cujomalainey** [comment]
> > Another point raised, why do we have a blob size restriction in the adapter? I thought we have size restriction tools in the topology, shouldnt we use that rather than create duplicate code?
> > [1ac1adb](https://github.com/thesofproject/sof/commit/1ac1adb5b1a0e7f942216d26b471547e1dae1de8)
> 
> Can't think of any FW/SW reason, could be a limitation due to available memory in the initial implementation target ?

The limit is set way below any potentially dangerous range. On Kpb we are using 10x the limit. Just seems redundant.

**Comment by mrajwa** [comment]
> you appear to have a double check, one in [generic](https://github.com/thesofproject/sof/blob/master/src/audio/codec_adapter/codec/generic.c#L285) and another [codec_adapter](https://github.com/thesofproject/sof/blob/master/src/audio/codec_adapter/codec_adapter.c#L537) and the codec_adapter check does not emit a message and comes first.

It isn't double check, the codec_adapter part checks the state of CODEC_ADPTER `if (cd->state >= PP_STATE_PREPARED)` while the generic part checks state of the CODEC `if (cd->codec.state < CODEC_PREPARED)` don'y you see the difference?
Also you say 

> the codec_adapter check does not emit a message and comes first.

Doesn't emit any message because its not considered an error situation - the config has been copied but not applied because codec had not been ready for that. Maybe we can add a warning or info log there.


> I think IIR and FIR components are an excellent example where they can take a config at anytime and just pick it up when its time to use it.

But CODEC_ADAPTER does the same for RUNTIME params. It just stores them in memory and passes to codec at time it is ready (in proper state).


> The limit is set way below any potentially dangerous range. On Kpb we are using 10x the limit. Just seems redundant.

How do you define that dangerous range? For the FIR component you mentioned earlier -> https://github.com/thesofproject/sof/blob/master/src/audio/eq_fir/eq_fir.c#L333
The blob size limit has several reasons:
- we don't want to send large blobs configs to avoid high load spikes hence the limit was set to 4k per cAVS documentation
- the approximated size of runtime config was calculated on the reference codec to fit in that area
- to avoid FW attack - as you pointed out the limit is in the topology so anybody can remove the topology limit, push large blob of a trash even and cause DSP run out of memory...

Anyway the limit can be increased by integrators as they which with the care of course :)

**Comment by cujomalainey** [comment]
> > you appear to have a double check, one in [generic](https://github.com/thesofproject/sof/blob/master/src/audio/codec_adapter/codec/generic.c#L285) and another [codec_adapter](https://github.com/thesofproject/sof/blob/master/src/audio/codec_adapter/codec_adapter.c#L537) and the codec_adapter check does not emit a message and comes first.
> 
> It isn't double check, the codec_adapter part checks the state of CODEC_ADPTER `if (cd->state >= PP_STATE_PREPARED)` while the generic part checks state of the CODEC `if (cd->codec.state < CODEC_PREPARED)` don't you see the difference?
> Also you say
> 
> > the codec_adapter check does not emit a message and comes first.
> 
> Doesn't emit any message because its not considered an error situation - the config has been copied but not applied because codec had not been ready for that. Maybe we can add a warning or info log there.
> 

Based on what I am reading in the code this means that the codec adapter is managing (essentially) the same state machine 3 times over (versus most components only managing 1). Why do we need an additional 2 that essentially mirror the component state machine? Why are we not inferring (and directly basing) our state off the component state? I can see places in the code where the state machines can become desynced (e.g. prepare) if the codec calls fail but the component state is already advanced. This additional managerial code seems unnecessary and adding potential additional error paths that could be simplified.

> > I think IIR and FIR components are an excellent example where they can take a config at anytime and just pick it up when its time to use it.
> 
> But CODEC_ADAPTER does the same for RUNTIME params. It just stores them in memory and passes to codec at time it is ready (in proper state).

The difference is that the component is notified. The codec should make the choice to defer, not enforce that in the code_adapter. If someone writes a codec that takes a long time to config, users may want to set it up before prepare to not further delay pcm open time.

> 
> > The limit is set way below any potentially dangerous range. On Kpb we are using 10x the limit. Just seems redundant.
> 
> How do you define that dangerous range? For the FIR component you mentioned earlier -> https://github.com/thesofproject/sof/blob/master/src/audio/eq_fir/eq_fir.c#L333
> The blob size limit has several reasons:
> 
> * we don't want to send large blobs configs to avoid high load spikes hence the limit was set to 4k per cAVS documentation

Intel is not the only one who uses this code, therefore restrictions per platform should be inferred based on the selected platform, not reducing the code to the lowest common denominator.

> * the approximated size of runtime config was calculated on the reference codec to fit in that area

A single data point is not a generalized case

> * to avoid FW attack - as you pointed out the limit is in the topology so anybody can remove the topology limit, push large blob of a trash even and cause DSP run out of memory...

I agree we should have a strong defense against the kernel which is why I am leading the fuzzing effort. But we should handle memory correctly and report OOM issues rather than just banning the use case blindly. If we want a firmware side check then I would argue we should build this into the API and have the codec dictate a sane size for its config to protect against these kind of attacks. This will allow finer controls and also not prevent slightly larger configs from being instantiated.

> 
> Anyway the limit can be increased by integrators to match their needs....but in the same time being aware of the potential danger I mentioned above.

**Comment by mrajwa** [comment]
> The difference is that the component is notified. The codec should make the choice to defer, not enforce that in the code_adapter. If someone writes a codec that takes a long time to config, users may want to set it up before prepare to not further delay pcm open time.
> 
> > >

The design of codec adapter assumes sequential logic so every action starts with codec_adapter (the parent if you will) and then propagates down to codecs (its childs). So yes the codec_adapter is "managerial" in the sense that if for example .prepare() method has failed on its level (because we run out of memory for example) why would codec_adapter let codec prepare itself? In the same way components on the pipeline work - if we have a pipeline HOST -> VOLUME -> DAI and if the VOLUME component failed to prepare itself why would be proceed with DAI preparation? 
You propose overtake actions like (paraphrasing here) "codec should be able to apply its config, prepare itself no matter SoF component (codec_adapter) state" I am afraid this approach will lead to more serious issues but anyway - how would you like to trigger codec preparation? Right now it is part of SoF component .prepare() method, would you like to send additional IPC?

> I can see places in the code where the state machines can become desynced (e.g. prepare) if the codec calls fail but the component state is already advanced.

Where do you see such thing in the .prepare()? On the codec_adpater level the state change of SoF component is done as the very last thing of .prepare() `cd->state = PP_STATE_PREPARED;` on the generic part, which changes state of codec similarly, very last thing `codec->state = CODEC_PREPARED;` so please don't make up issues.

> codec adapter is managing (essentially) the same state machine 3 times over (versus most components only managing 1). 

No, it is not the same state machine, codec_adapter during .prepare() method performs all the checks important from SoF component perspective like "do we have sources & sinks, assigned, do we have proper state, memory available once this is verified it ask codec to prepare itself with `ret = codec_prepare(dev);`  defined in the glue code codec_adapter/generic.c. Next that part barely checks if the codec has already been prepared with `if (cd->codec.state == CODEC_PREPARED) return 0;` and if not immediately calls codec specific prepare by `ret = codec->ops->prepare(dev);`. Having said that I don't see what three state machines you are talking about. I agree that from some perspective this generic part may seem redundant, like in the case of prepare it doesn;t do much and its tempting to get rid of that and call codec interfaces right from the adapter part. Sure it can be done but will obfuscate the SoF component operations with the codec ones - inspect the generic .init() method for example. Also currently we do support only one the cadence codec once more codecs will be added the generic part may have more logic so I believe this separate layer will turn out to be beneficial.

> Intel is not the only one who uses this code, therefore restrictions per platform should be inferred based on the selected platform, not reducing the code to the lowest common denominator.

Sure that's why I mentioned integrator who can increase this one line limit to match their needs. Btw I think we should overload this macro with codec specific limit so then no change in the code will be needed.

**Comment by cujomalainey** [comment]
> > The difference is that the component is notified. The codec should make the choice to defer, not enforce that in the code_adapter. If someone writes a codec that takes a long time to config, users may want to set it up before prepare to not further delay pcm open time.
> > > >
> 
> The design of codec adapter assumes sequential logic so every action starts with codec_adapter (the parent if you will) and then propagates down to codecs (its childs). So yes the codec_adapter is "managerial" in the sense that if for example .prepare() method has failed on its level (because we run out of memory for example) why would codec_adapter let codec prepare itself? In the same way components on the pipeline work - if we have a pipeline HOST -> VOLUME -> DAI and if the VOLUME component failed to prepare itself why would be proceed with DAI preparation?
> You propose overtake actions like (paraphrasing here) "codec should be able to apply its config, prepare itself no matter SoF component (codec_adapter) state" I am afraid this approach will lead to more serious issues but anyway - how would you like to trigger codec preparation? Right now it is part of SoF component .prepare() method, would you like to send additional IPC?

Error cases are different than use cases. Clearly yes the adapter should simplify error handling at the component level. The codec should be able to be configured whenever, and be able to self identify if it is configured. If it is not configured at the prepare step and no data was presented with prepare then yes we should fail. I am saying we should decouple the configuration and the prepare step as the component model already does. Right now components like IIR/DCblock,etc can be reconfigured at anytime, unprepared or during playback. I don't understand why you are taking away that freedom in the adapter when clearly this model already works in the rest of SOF. There are no need for additional IPC, you just need not prematurely return and let the config call make it all the way to the codec.

> 
> > I can see places in the code where the state machines can become desynced (e.g. prepare) if the codec calls fail but the component state is already advanced.
> 
> Where do you see such thing in the .prepare()? On the codec_adpater level the state change of SoF component is done as the very last thing of .prepare() `cd->state = PP_STATE_PREPARED;` on the generic part, which changes state of codec similarly, very last thing `codec->state = CODEC_PREPARED;` so please don't make up issues.

As I said, the issue is not between your two internal states, its between the component state and your internal states. You set the component state [here](https://github.com/thesofproject/sof/blob/master/src/audio/codec_adapter/codec_adapter.c#L206) but you have [three](https://github.com/thesofproject/sof/blob/master/src/audio/codec_adapter/codec_adapter.c#L221) [different](https://github.com/thesofproject/sof/blob/master/src/audio/codec_adapter/codec_adapter.c#L254) [error](https://github.com/thesofproject/sof/blob/master/src/audio/codec_adapter/codec_adapter.c#L261) cases after you set the state. You fail to walk back the state to RESET as is [practice](https://github.com/thesofproject/sof/blob/master/src/audio/volume/volume.c#L895). The last two actually happen after you set the codec state, which could leave you in a state of ` comp=PREPARE, comp_data=CREATED, codec=PREPARE`. I understand this criticism is frustrating for you, but simply because you do not see it does not give you a right to accuse me of false fabrication.

> 
> > codec adapter is managing (essentially) the same state machine 3 times over (versus most components only managing 1).
> 
> No, it is not the same state machine, codec_adapter during .prepare() method performs all the checks important from SoF component perspective like "do we have sources & sinks, assigned, do we have proper state, memory available once this is verified it ask codec to prepare itself with `ret = codec_prepare(dev);` defined in the glue code codec_adapter/generic.c. Next that part barely checks if the codec has already been prepared with `if (cd->codec.state == CODEC_PREPARED) return 0;` and if not immediately calls codec specific prepare by `ret = codec->ops->prepare(dev);`. Having said that I don't see what three state machines you are talking about. I agree that from some perspective this generic part may seem redundant, like in the case of prepare it doesn;t do much and its tempting to get rid of that and call codec interfaces right from the adapter part. Sure it can be done but will obfuscate the SoF component operations with the codec ones - inspect the generic .init() method for example. Plus all the codec_related defines, functions, helpers etc are kept out of codec_adapter. Currently we do support only one the cadence codec, once more codecs will be added the generic part may have more logic so I believe this separate layer will turn out to be beneficial.

I dont understand why you need these. I understand what they represent, but given most of these calls are linear (e.g. prepare in codec is only called by prepare in generic, which is only called by prepare in codec_adapter prepare) the states are in lockstep with each other. Unless there are plans to leave the codec in different states than the component, this additional state information (cd->state, codec->state) is superfluous and leads to more possible states and complexity. Assuming the codec will continue to follow in step with the component you can write simple helper functions to return simplified states. e.g. 

```
state is_codec_running(comp_dev *dev) {
return dev->state == COMP_STATE_ACTIVE || dev->state==COMP_STATE_PAUSED;
}
```

This reduces the possible state space by keeping a single source of truth. Please share your future plans for these states otherwise they appear to be unnecessary additional complexity.

> 
> > Intel is not the only one who uses this code, therefore restrictions per platform should be inferred based on the selected platform, not reducing the code to the lowest common denominator.
> 
> Sure that's why I mentioned integrator who can increase this one line limit to match their needs. Btw I think we should overload this macro with codec specific limit so then no change in the code will be needed.

That just leads endless forks, even among intel platforms, we have the tools to configure based on per board basis, why not use them?

**Comment by mrajwa** [comment]
> I understand this criticism is frustrating for you, but simply because you do not see it does not give you a right to accuse me of false fabrication.

Actually it is the opposite, I appreciate your criticism and see it as the opportunity to improve the code as well as my thinking maybe. I now understand better what you had in mind in regards to states but I still feel like you make up things, here's another example:

> Right now components like IIR/DCblock,etc can be reconfigured at anytime, unprepared or during playback

Can you prove IIR can be "reconfigured unprepared"? I will prove it can not -> if IIR is unprepared the new config can ONLY be copied into its internal memory with the help of `comp_data_blob_set_cmd` [here,](https://github.com/thesofproject/sof/blob/master/src/audio/eq_iir/eq_iir.c#L655) the actual reconfiguration happens either during .prepare() [here](https://github.com/thesofproject/sof/blob/master/src/audio/eq_iir/eq_iir.c#L795) or during .copy() [here](https://github.com/thesofproject/sof/blob/master/src/audio/eq_iir/eq_iir.c#L734). End of the proof &#8718;
BTW, the later solution to reconfigure during .copy() is a bad idea from performance perspective. codec_adapter does a much better job here since it applies new config at the time it receives it which guarantees the processing latency is not affected.

**Comment by mrajwa** [comment]
> You fail to walk back the state to RESET as is [practice](https://github.com/thesofproject/sof/blob/master/src/audio/volume/volume.c#L895).

Technically yes, the error handling of prepare could reset the component but in practice it doesn't matter because what follows after failed prepare is globack comp_reset() which will do the same :)

**Comment by cujomalainey** [comment]
> > Right now components like IIR/DCblock,etc can be reconfigured at anytime, unprepared or during playback
> 
> Can you prove IIR can be "reconfigured unprepared"? I will prove it can not -> if IIR is unprepared the new config can ONLY be copied into its internal memory with the help of `comp_data_blob_set_cmd` [here,](https://github.com/thesofproject/sof/blob/master/src/audio/eq_iir/eq_iir.c#L655) the actual reconfiguration happens either during .prepare() [here](https://github.com/thesofproject/sof/blob/master/src/audio/eq_iir/eq_iir.c#L795) or during .copy() [here](https://github.com/thesofproject/sof/blob/master/src/audio/eq_iir/eq_iir.c#L734). End of the proof ∎

It appears I am referencing the old implementation pre 375f84473f2e2b57d72ad91454b0832b110ed3ff But that doesn't change my point. The component gets a callback at any point and handles the config call when it chooses to. The codec adapter is not passing through the callback unless it is in certain states. I just don't understand the design logic behind restricting when the component can choose to act upon its config.

> BTW, the later solution to reconfigure during .copy() is a bad idea from performance perspective. codec_adapter does a better job here since it applies new config at the time it receives it (assuming proper state) which guarantees the processing latency is not affected.

Agreed, but why not configure when there are no chance of latency issues at all (i.e. before prepare)? I agree that changing config in copy is not great but if i remember correctly, the reason here is to make sure that the change happens between periods and does end up with a config change mid process as the config changes are non-atomic. Makes me wonder if we should associate a configure callback with model handler, and handler checks dev state itself and calls directly if we are no running, otherwise calls right before copy callback. Might reduce some boilerplate code.

> 
> > You fail to walk back the state to RESET as is [practice](https://github.com/thesofproject/sof/blob/master/src/audio/volume/volume.c#L895).
> 
> Technically yes, the error handling of prepare could reset the component but in practice it doesn't matter because what follows after failed prepare is globack comp_reset() which will do the same :)

This is not always true, it is true in IPC context, but it appears we are lacking that logic in the [XRUN triggers](https://github.com/thesofproject/sof/blob/master/src/audio/pipeline.c#L803). I agree, not having the walkback logic would make for cleaner callbacks, so I would argue we should look at the XRUN path and fix it.

**Comment by mrajwa** [comment]
> It appears I am referencing the old implementation pre [375f844](https://github.com/thesofproject/sof/commit/375f84473f2e2b57d72ad91454b0832b110ed3ff) But that doesn't change my point. The component gets a callback at any point and handles the config call when it chooses to. The codec adapter is not passing through the callback unless it is in certain states. I just don't understand the design logic behind restricting when the component can choose to act upon its config.
> 
> >

How about changing the condition which currently passes runtime config only when codec is prepared to codec initialized? 
I believe at the very minimum we should be sure codec instance was created.

**Comment by cujomalainey** [comment]
> > It appears I am referencing the old implementation pre [375f844](https://github.com/thesofproject/sof/commit/375f84473f2e2b57d72ad91454b0832b110ed3ff) But that doesn't change my point. The component gets a callback at any point and handles the config call when it chooses to. The codec adapter is not passing through the callback unless it is in certain states. I just don't understand the design logic behind restricting when the component can choose to act upon its config.
> > 
> > >
> 
> How about changing the condition which currently passes runtime config only when codec is prepared to codec initialized? 
> I believe at the very minimum we should be sure codec instance was created.

Sounds good to me.

**Comment by cujomalainey** [comment]
@lgirdwood i think we can go ahead and merge this, the rest of the topics are to be addressed in the meeting
